core/trie2/trienode/node.go

```diff
@@ -0,0 +1,141 @@
+package trienode
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/NethermindEth/juno/core/crypto"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/core/trie2/trieutils"
+)
+
+var (
+	_ Node = (*BinaryNode)(nil)
+	_ Node = (*EdgeNode)(nil)
+	_ Node = (*HashNode)(nil)
+	_ Node = (*ValueNode)(nil)
+)
+
+type Node interface {
+	Hash(crypto.HashFn) felt.Felt
+	Cache() (*HashNode, bool)
+	Write(*bytes.Buffer) error
+	String() string
+}
+
+type (
+	// Represents a binary branch node in the trie with two children
+	BinaryNode struct {
+		Children [2]Node // 0 = left, 1 = right
+		Flags    NodeFlag
+	}
+	// Represents a path-compressed node that stores a path segment
+	// and a single child node
+	EdgeNode struct {
+		Child Node            // The child node at the end of the path
+		Path  *trieutils.Path // The compressed path segment
+		Flags NodeFlag
+	}
+	// Represents a node that only contains a hash reference to another node
+	HashNode struct{ felt.Felt }
+	// Represents a leaf node that stores an actual value in the trie
+	ValueNode struct{ felt.Felt }
+)
+
+// Used when collapsing internal trie nodes for hashing, since unset children need to be hashed correctly
+var NilValueNode = &ValueNode{felt.Felt{}}
+
+type NodeFlag struct {
+	Hash  *HashNode // The cached hash of the node
+	Dirty bool      // Whether the node has been modified
+}
```

Comment:
Checking the PR there is no need for this type to be public, should it be?

Reply:
Why?

Reply:
imprecise, I meant there is no need for it to be public